### ISSUE TITLE
Question: lizardfs on a single machine

### ISSUE THREAD

**Issue by eleaner** [lizardfs/lizardfs#764]
Hi,

I am running a backup machine at home with a bunch(fifteen) of different HDD. From brand new 10TB to old 1TB. The machine runs only for a few hours a day until backups are done.
I am using BTRFS in so-called RAID. With BTRFS it guarantees that each file is stored on two physically different drives. BTRFS also handles different sizes of the drives well.

The idea is to recycle drives from other machines. Every old drive that is bigger than the smallest in the pool could increase my storage and be useful for some more time. The one that would drop out of the pool was, a long time ago, meant for bin anyway. 

Recently I started using lizardfs for my network backups and I love it.
It made me think - what if I replace BTRFS on my backup machine with lizardfs? Using EC I could even gain some storage space.
so here are the questions
1. I could run master and chunkserver on the same machine. I could add 15 drives to the chunkserver. Would this work? Or would I have to run 15 virtual chunk servers?
2. how would EC behave with so different disk sizes? What would mean if I set EC on 10,5? Randomly peaking from my pool five drives could have less than 10TB of space or as much as 30TB.

**Comment by onlyjob** [comment]
Chunkserver can easily handle 15+ HDDs on the same machine with Master. The only constraint is RAM.
However one chunkserver can only hold one replica. For EC goals you need many chunkservers and I've found EC goals of 6+ chunks impractical...

**Comment by eleaner** [comment]
thank you @onlyjob

So instead of starting 15 chunk servers should be better to start e.g five in ecc 4,1
Then make sure that the sum of the disk space for each chunk server is comparable to the others.
Am I correct?

**Comment by njhurst** [comment]
I have two systems run exactly as you propose, @eleaner,  I'm using {5,3} with a separate CS per drive, > 15 drives.  It all works fine, I'm not sure what aspect @onlyjob finds impractical though.  Write a script to setup the chunkservers in a systematic way, and use ext4 not btrfs for each drive.  I also came from a btrfs machine with many drives in 'raid1', but lizard is much easier to adjust and monitor.

**Comment by eleaner** [comment]
Thank you @njhurst 
Do you know how to estimate the memory and CPU requirements?
My backup machine is rocking G3220T @ 2.60GHz, and 16GB ram. Ideally, I would like to stick to that?
I would like to drop s3ql on top of that as a backup+deduplication filesystem so I might be short on resources

**Comment by njhurst** [comment]
My systems are 16 core somethings with hundreds of GB of ram... but the cpu usage is < 50% of 1 core under normal operation, ram is 3GB.  This is for a 100TB system with around 10M chunks.  I think your specs are plenty.

**Comment by 19wolf** [comment]
I'm doing something similar, and wrote a script to make setup easier. Here it is

```
#!/bin/bash

IPADDR=$(ip route get 1.1.1.1 | awk '/1.1.1.1/ {print $7}')
PORT=0

for DISKDIR in $(ls /mnt/); do

  if mountpoint /mnt/$DISKDIR; then

      mkdir /var/lib/mfs/chunkserver-$DISKDIR;
      #rm /mnt/$DISKDIR/.lock;

      chown -R mfs:mfs /mnt/$DISKDIR;
      chown -R mfs:mfs /var/lib/mfs/chunkserver-$DISKDIR;

      echo "MASTER_HOST =" $IPADDR > /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "SYSLOG_IDENT = chunkserver-$DISKDIR" >> /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "CSSERV_LISTEN_PORT = $(( 9500 + $PORT))" >> /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "DATA_PATH = /var/lib/mfs/chunkserver-$DISKDIR" >> /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "HDD_CONF_FILENAME = /etc/mfs/mfshdd-$DISKDIR.cfg" >> /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "HDD_LEAVE_SPACE_DEFAULT = 50GiB" >> /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      echo "/mnt/$DISKDIR" > /etc/mfs/mfshdd-$DISKDIR.cfg;

      chown mfs:mfs /etc/mfs/mfschunkserver-$DISKDIR.cfg;
      chown mfs:mfs /etc/mfs/mfshdd-$DISKDIR.cfg;
      echo "Staring chunkserver $PORT"

      (( PORT++ ))

      mfschunkserver -c /etc/mfs/mfschunkserver-$DISKDIR.cfg; echo "Started $DISKDIR: $?";

  fi
done
```

**Comment by onlyjob** [comment]
Just some comments:

1) Don't use _ec(4,1)_ or any EC goals with just one redundancy chunk (,1). This is RAID-5 level of "safety" which means _unsafe_ hence it is only for disposable data i.e. "Trash", old backups and everything which is OK to delete right now.

2) EC(\*,1) performs slower than equivalent _xor_ chunks so it is better to use _xor_ goals for all disposable data. Let me stress it again that _ec(\*,1)_ and _xor_ goals are only for disposable data because of RAID-5 level or redundancy.

3) _ec(\*,2)_, _ec(\*,3)_ are good goals except that upgrade to 3.13.0~rc1 _destroys them_ (#746). Reason why I called _ec_ goals with 6+ splits impractical is because LizardFS can only have one chunk split per chunkserver hence changing goal from _ec(2,2)_ to _ec(3,2)_ would require 9 chunkservers. Too many chunkservers have overhead to resources and to systems administration. Ideal LizardFS setup is many HDDs per chunkserver; (no more than) few chunkservers per server.

4) Do yourself a favour and use [ansible](https://en.wikipedia.org/wiki/Ansible_(software)) instead of bash scripts for automation and systems administration. :)

**Comment by eleaner** [comment]
hi @onlyjob 

> LizardFS can only have one chunk split per chunkserver hence changing goal from ec(2,2) to ec(3,2) requires 9 chunkservers

I am sorry but I don't understand that. Please tell me where my reasoning is wrong
ec(2,2) = 2 chunks with data and 2 chunks parity = minimum 4 chunkservers
ec(3,2) = 3 chunks with data and 2 chunks parity = minimum 5 chunkservers

In can come up with number 9 only if I sum up those two numbers. Why would I do that?
Do you say that when **changing goals** you need a full suite of new servers for a new goal and it cannot be done in place?

**Comment by onlyjob** [comment]
@eleaner, I think you understood correctly. :) LizardFS always make a chunk replica on another chunkserver that doesn't have it already. That include changing goals as well.
This is actually a logical thing: for safety you can't have more than one copy of a chunk per chunkserver or goal definition will be violated. Not sure how to express it better...

Chunks are not converted in-place when the goal is changing. That would be difficult or impossible to do. When you change a goal LizardFS makes a new set of chunks for a new goal then removes old chunks. That requires up to 200% of space - more precisely whatever space take the old goal plus space required by new goal. Massive goal change can consume (a lot of) free space in your cluster.

**Comment by eleaner** [comment]
@onlyjob 
I understand the space requirements, but I still don't understand why I would need extra chunk servers.
I am pretty sure that one chunk server can hold two chunks from the same file but different goals. Or?

**Comment by onlyjob** [comment]
I don't know how to explain then... I've observed this behaviour by trying to change goal and I've noticed that goal change stuck without enough chunkservers. Try changing a goal of a file and observe how chunks are created using `lizardfs fileinfo`...

**Comment by eleaner** [comment]
I think I understand. You say that if have ec2,2 on 4 chunkservers and I want to add chunkserver and move to ec3,2 I will need extra 5 chunkservers. At least for the time of conversion.
I get it, but if it is true it's a major drawback of my idea of single machine setup (or probably any setup).
I think i will start another thread and give chance to someone from dev team to answer

**Comment by onlyjob** [comment]
Yes, for the time of conversion. Possible workaround it convert goal to intermediate _std_ then back to _ec_. It is not a big limitation but definitely something to keep in mind. Single machine setup is an unusual idea. MooseFS and LizardFS are designed with chunkserver (server) as a unit of failure. Although stacking many chunkservers per server allows to have space efficient EC goals, it does not particularly help reliability as your data will be inaccessible when machine with many chunkservers is down.

**Comment by onlyjob** [comment]
Also you can _copy_ data in order to change goal - this allows to bypass limitation for number of chunkservers.

**Comment by onlyjob** [comment]
No worries. Almost 4 years of experience operating LizardFS cluster pays off... :)

When your smallest chunkserver is filled there will be nowhere to create chunks to satisfy EC goal (if you only have 6 chunkservers for `ec(3,3)` goal).
I think first you'll see undergoal chunks, then (when more chunkservers run out of space) you won't be able to save files...

Replicated goals are balanced perfectly. LizardFS will make replicas as required as long as you have space anywhere in the cluster, placing no more than one replica per chunkserver. This works really well with mixed HDD sizes and different chunkservers capacities.

Changing goal from `$std { _ _ }` to `$std { _ }` needs no extra chunkservers as LizardFS just have to remove redundant chunk copies.

Replicated goals are straightforward and easy to understand. Almost always they just work as expected.

As far as I understand, `std1 --> xor3` needs 5 chunkservers...

**Comment by njhurst** [comment]
comments:
1. lizard seems much easier to change than btrfs - every btrfs operation tends to go into a locked kernel state which is unpausable, unmonitorable and can fail catastrophically.  I had many more problems with a single machine btrfs than I've ever had with a single machine lizardfs.  But I haven't tried btrfs for a few years now, maybe it's fixed.  Having real FEC is worth the hassle too.  I also like that I can just move chunk data around with normal tools (e.g. rsync) and my data storage is independent of my disk format.

1.5. Don't forget that you can easily generate recovery backups of metadata for lizard, which is impossible AFAIK for btrfs.  I cron-rsync my metadata to another place and I've successfully recovered by just rsyncing it back.

2. mathematically there's no reason why going from {2,2} to {2,2+k} can't just generate the extra k checksum blocks (not true if you change the identity blocks though, so 2,2 to 3,2 will require a complete resum).  each checksum block is orthogonal, so we can always generate a new orthogonal block up to the dimensionality of the blocks. likewise dropping an arbitrary checksum block from say {2,3} to {2,2} is perfectly ok without any resumming.
3. there's nothing unsafe about having a {2,2} and {3,2} copy using the same chunkservers, as long as you accept that you only have the safety of the original goal until the new goal is reached.  Whether lizardfs takes this tack I don't know, let's see if anyone answer's @eleaner 's question.
4. interesting but probably never implemented: You can go from {n,m} to {kn,km} almost in-place, then as per 2. adjust up or down in place.

5. regarding your chunkservers and hdds, I would make each chunkserver have about the same space.  This is easy to do - turn off the system (halt master, cs etc), move the directories around in your mfshdd.cfg files, turn it all back on.  It will just magically have the new allocation.  So perhaps [(10,), (10,), (7,4), (7,4)].  You can also run more than one CS per drive - just make two directories.  The allocation algorithm seems reasonable stable in this case as it keeps a margin open (`HDD_LEAVE_SPACE_DEFAULT = 4GiB`), as long as the margin is larger than the 'loop gain' (GB/policy update).  But of course you will have to do your risk analysis with the assumption that any single failure will take out 2 CS.  This can still be ahead space and risk wise.

**Comment by eleaner** [comment]
thanks @njhurst 
I was thinking an option to go ec2,1 on my configuration with
2x10TB + 2x7TB + 2x4TB
I am not sure if algorithms are clever enough but in theory, it should balance nicely down to about 2TB space that cannot be safely filled with this goal.

but I start liking your idea of going [(10,), (10,), (7,4), (7,4)] with ec3.1

@njhurst and @onlyjob - you guys are rockstars

**Comment by njhurst** [comment]
Thanks @eleaner I agree that @onlyjob is a rockstar :)  They also made the point that if you want ec{n, 1} you perhaps should use xor, which is perhaps a bit faster.

**Comment by eleaner** [comment]
valid point :(
ec{n,1} will blow if when one HDD is down, one other chunk is deteriorated. Just like in-famous raid5. And with 10T probability of URE is pretty high on my HDDs.
So ec{2,2} then, way more secure than my last setup (which was ZFS based strip of three mirrors) setups.

**Comment by njhurst** [comment]
Actually, there is a difference between xor goal and raid5 - raid5 will become corrupted silently because it doesn't have checksums, whereas lizard and btrfs will correctly identify the bad block/chunk in the continuous check loop.  So losing a drive with more than n+m chunkservers it will just immediately start backfilling those chunks onto other drives.  Which is why you really want CS = n+m+k, with k being your margin for emergency drive buying.  You have to think about how important your data is, there really is no way to protect data other than to have lots of drives, if this is not economical for you, and the redundancy is required, consider using an online backup service such as backblaze.  No fancy algorithm can work around the fact that the most common failure mode these days is a whole drive becoming unreadable without warning.  Given that fact, I don't know of any data protection system that is qualitatively better than the approach lizard uses.
Personally I stopped buying big drives, as the overhead for smaller drives turned out to be less important than the much lower performance I've measured from the biggest drives.  I stick with 4TB drives and just use lots of them, the $/TB is actually better at the point too.  Second hand drive bay servers are dirt cheap on ebay these days ($200 will get you a machine with your specs and 12 drive bays in 2U).

**Comment by njhurst** [comment]
hmmm, yes, I think you are right that it is always safer using the maximum factor available for your drives.  It would depend on whether lizard prioritises by missing chunk ratio or not I guess.

**Comment by njhurst** [comment]
look on ebay for xeon server.

**Comment by 19wolf** [comment]
I've seen plenty of cheap servers on ebay, I was hoping you could point me to a specific model?
#offtopic

**Comment by njhurst** [comment]
I use 12bay 2U supermicros, but I can't say whether they are especially good or bad.  Been working solidly for 3 or 4 years.

**Comment by njhurst** [comment]
hmmm, yes, I think you are right that it is always safer using the maximum factor available for your drives.  It would depend on whether lizard prioritises by missing chunk ratio or not I guess.

**Comment by eleaner** [comment]
@19wolf 
My backup is 95% my photos. Other things are not even worth mentioning. I was just trying to understand whether std{ _ _ } or ec{2,2} and I decided for the second as I believe it is far superior. With equally sized HDD I would probably go ec{4,2}

**Comment by 19wolf** [comment]
It seems you could accomplish the same thing using std{_ _ _ _} as you do with ec{2,2}, no? They both have 400% overhead, if my [calculations](https://www.reddit.com/r/DataHoarder/comments/99ubi0/erasure_coding_overhead/e4qma8e/) are correct.

**Comment by eleaner** [comment]
I believe std{_ _ _ _} is 300% overhead, 4x storage requirements
ec{2,2} is two data chunks and two parity chunks = 100% overhead, 2x storage requirements.

**Comment by 19wolf** [comment]
That makes sense. I haven't slept in a while...

Why did you decide ec{2,2} over std{_ _}, given they both have 2x storage requirement? ec{2,2} seems like it just adds cpu time to calculate parity.

**Comment by eleaner** [comment]
ec{2,2} will allow two chunk servers to go down "safely"
std{_ _} just one

**Comment by eleaner** [comment]
Would lizardfs benefit from the use of a few smaller SSD (in a form of a cache?) in such a single machine cluster?

3xSSD 240GB
9HDD - 30TB in total

**Comment by Zorlin** [comment]
It should do. You could use something like bcache or enhanceIO. I believe
Cloudweavers/nodeweavers use a slightly modified version of enhanceIO in
their stack.

**Comment by eleaner** [comment]
while bcache says something to me - it is not much
do you mean caching drives before exposing to lizardfs?
or something else?

**Comment by Zorlin** [comment]
Yes, I mean putting a cache in front of your disks and then exposing that to LizardFS.

**Comment by pasqd** [comment]
Hi,
Im testing ZFS pool caching for my chunkservers on lizardfs ec(6+2).
I will post the results once i finish :)

**Comment by aminhusni** [comment]
Hi, 

Since there is scarce information on how to deploy LizardFS on a single machine with multiple drives as the chunk servers, I've taken the time to write a tutorial here for anybody who want to try it out:

[Tutorial] LizardFS – Drive Level File Redundancy On Single Computer
https://nasilemaktech.com/guide-lizardfs-drive-level-redundancy/

**Comment by Blackpaw** [comment]
Thank you very much, that's great.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
